# Линейная регрессия. Работа с признаками

## Описание задачи и загрузка данных

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

В этом задании мы рассмотрим различные аспекты построения линейной модели. Мы будем работать с одним из классических наборов данных в статистике, содержащим информацию о бриллиантах. Описание можно посмотреть [здесь](https://www.kaggle.com/shivam2503/diamonds).

In [ ]:
data = pd.read_csv('https://raw.githubusercontent.com/evgpat/edu_stepik_practical_ml/main/datasets/diamonds.csv')
data.head(5)

,Unnamed: 0,carat,cut,color,clarity,depth,table,price,x,y,z
0,1,0.23,Ideal,E,SI2,61.5,55.0,326,3.95,3.98,2.43
1,2,0.21,Premium,E,SI1,59.8,61.0,326,3.89,3.84,2.31
2,3,0.23,Good,E,VS1,56.9,65.0,327,4.05,4.07,2.31
3,4,0.29,Premium,I,VS2,62.4,58.0,334,4.20,4.23,2.63
4,5,0.31,Good,J,SI2,63.3,58.0,335,4.34,4.35,2.75


Посмотрим на типы столбцов.

In [ ]:
data.dtypes

,0
Unnamed: 0,int64
carat,float64
cut,object
color,object
clarity,object
depth,float64
table,float64
price,int64
x,float64
y,float64


Мы будем решать задачу предсказания цены бриллианта `price` в зависимости от его характеристик.

## Построение модели

### Задание 1

Есть ли в наборе данных пропущенные значения? Если да, удалите их.

Также выведите на экран число пропусков в каждом столбце.

In [ ]:
# your code here
data.isnull().sum()

,0
Unnamed: 0,0
carat,0
cut,0
color,0
clarity,0
depth,0
table,0
price,0
x,0
y,0


Ответ: 0

### Задача 2

Есть ли в наборе данных бессмысленные столбцы (признаки, не несущие дополнительной информации)?  
Если да, то удалите их.

In [ ]:
# your code here
data.drop(columns=['Unnamed: 0'], inplace=True)

Ответ: Удалил `Unnamed: 0`

### Задание 3

Линейная регрессия основана на предположении о линейной связи между признаками и целевой переменной, а потому перед выбором переменных для включения в модель имеет смысл проверить, насколько эта связь выполняется. Для следующих пунктов нам также потребуются корреляции между признаками.  
Выведите матрицу корреляций между всеми вещественными признаками и целевой переменной.

Какой вещественный признак коррелирует с целевой переменной больше всего?

In [ ]:
# your code here
data_num = data.select_dtypes(include=['float64'])
data_num.corrwith(data['price']).sort_values(ascending=False)

,0
carat,0.921591
x,0.884435
y,0.865421
z,0.861249
table,0.127134
depth,-0.010647


Ответ: `carat`

### Задание 4

Так как линейная модель складывает значения признаков с некоторыми весами, нам нужно аккуратно обработать категориальные признаки. Закодируйте категориальные переменные при помощи OneHot-кодирования ([`pd.get_dummies`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html)). Не забудьте поставить значение параметра `drop_first` равным `True`.

Сколько получилось столбцов в таблице `data`?

*P.S. Числовые столбцы оставляем в таблице без изменений.*

In [ ]:
# your code here
data = pd.get_dummies(data, drop_first=True)
data.shape[1]

24

Ответ: 24

### Задание 5

Создайте матрицу `X`, содержащую все признаки, и не содержащую целевую переменную `price`. Также создайте вектор `y`, содержащий целевую переменную `price`.

In [ ]:
# your code here
X = data.drop(columns=['price'])
y = data['price']

Разделите выборку на тренировочную и тестовую. Долю тестовой выборки укажите равной `0.3`.

При разбиении укажите `random_state = 42`.

In [ ]:
from sklearn.model_selection import train_test_split

Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, random_state = 42) # your code here

### Задание 6

Зачастую при использовании линейных моделей вещественные признаки масштабируются.  В этой задаче масштабируйте вещественные признаки тренировочной и тестовой выборок при помощи модуля `StandardScaler`.

*  Обучите (`fit`) scaler на тренировочных данных
*  Преобразуйте (`transform`) и трейн, и тест

После применения масштабирования матрица перестает быть объектом `pandas.DataFrame` - решите эту проблему.

In [ ]:
from sklearn.preprocessing import StandardScaler

# your code here
scaler = StandardScaler()

Xtrain = scaler.fit_transform(Xtrain)
Xtest = scaler.transform(Xtest)

Xtrain = pd.DataFrame(Xtrain, columns=X.columns)
Xtest = pd.DataFrame(Xtest, columns=X.columns)

# OR

# scaler = StandardScaler().set_output(transform="pandas")

# Xtrain = scaler.fit_transform(Xtrain)
# Xtest = scaler.transform(Xtest)

### Задание 7

Обучите линейную регрессию на тренировочной выборке. Выведите *r2-score* на тренировочной и тестовой выборках.

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# your code here
model = LinearRegression()
model.fit(Xtrain, ytrain)

print(r2_score(ytrain, model.predict(Xtrain)))
print(r2_score(ytest, model.predict(Xtest)))

0.9195976267987521
0.9201866914388087


Ответ: $r^2 \approx 0.92$

### Задание 8

Выведите на экран веса, которые линейная регрессия присвоила признакам.

Какой признак имеет наибольший отрицательный вес? (наибольший по модулю среди всех отрицательных весов)

In [ ]:
# your code here
weights = pd.Series(model.coef_, index=X.columns)
weights.sort_values(ascending=False)

,0
carat,5338.615671
clarity_VS2,1817.664208
clarity_VS1,1670.318496
clarity_SI1,1611.571681
clarity_VVS2,1461.034194
clarity_VVS1,1284.809475
clarity_SI2,1042.921215
clarity_IF,975.933641
cut_Ideal,414.369515
cut_Premium,339.239204


Ответ: `x`

## Попытка улучшить качество модели

Данная часть является творческой, и задания в ней немного сложнее, чем предыдущие.

Следующие вопросы не проверяются тестами.

### Задание 9

Как можно заметить из анализа корреляционной матрицы в задании 3, между некоторыми признаками имеется сильная корреляция, что может быть индикатором проблемы *мультиколлинеарности*. Различия в порядке коэффициентов, выявленные в предыдущей задаче, также свидетельствуют об этом. Для решения этой проблемы можно либо исключить некоторые признаки из модели (например, если признак линейно зависим с какими-то другими, его можно исключить из модели, т.е. удалить из матрицы объект-признак и заново обучить модель).

Удалите из матриц `Xtrain` и `Xtest` признак, который наиболее сильно коррелирует с остальными. Заново обучите модель и оцените её качество. Улучшилось ли качество модели?

Попробуйте удалить какой-то другой признак (можете попробовать несколько вариантов). Помогло ли это улучшить качество модели?

In [ ]:
# your code here
Xtrain.drop(columns=['x', 'y', 'z'], inplace=True)
Xtest.drop(columns=['x', 'y', 'z'], inplace=True)

model = LinearRegression()
model.fit(Xtrain, ytrain)

print(r2_score(ytrain, model.predict(Xtrain)))
print(r2_score(ytest, model.predict(Xtest)))

0.9159704730530022
0.916194716704243


Ответ: $r^2$ ухудшился с 0.92 до 0.916

### Задание 10

Иногда генерация новых признаков помогает модели лучше находить взаимосвязи между целевой переменной и признаками. Попробуйте придумать новые признаки и добавить их в модель.  

Помогло ли это улучшить качество модели?

In [ ]:
# your code here
X = X.assign(volume = lambda x: x['x'] * x['y'] * x['z'])
X = X.assign(density = lambda x: x['carat'] / x['volume'])
X = X.assign(depth_sq = lambda x: x['depth'] ** 2)
X = X.assign(table_sq = lambda x: x['table'] ** 2)

In [ ]:
# Use scikit-learn's SimpleImputer to replace NaNs with the median
from sklearn.impute import SimpleImputer

X = X.replace([np.inf, -np.inf], np.nan)

imputer = SimpleImputer(strategy='median')
X_clean = imputer.fit_transform(X)

In [ ]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X_clean, y, test_size=0.3, random_state = 42)

In [ ]:
scaler = StandardScaler().set_output(transform="pandas")

Xtrain = scaler.fit_transform(Xtrain)
Xtest = scaler.transform(Xtest)

In [ ]:
model.fit(Xtrain, ytrain)

print(r2_score(ytrain, model.predict(Xtrain)))
print(r2_score(ytest, model.predict(Xtest)))

0.9202601005073451
0.9203449639913093


Ответ: добавил volume (`x*y*z`), density(`carat`/`volume`) и квадраты формы пропорций (`dense`$^2$ и `table`$^2$). $r^2$ не особо сильно изменился, но все же увеличился, что значит, что новые переменные помогают улучшить качество модели.